### DISCUSSION TITLE
Notarization fails due to invalid folder names: Problem persists despite using previously-suggested fix.

### DISCUSSION THREAD

**Discussion by fishbacp** [pyinstaller/pyinstaller#7628] [Help] [answered]
I created my .app using Pyinstaller with a virtual environment as outlined in this [instructional video](https://www.youtube.com/watch?v=gsSjIx_uFG4). (Virtual environments are new to me.) My .spec file was configured to create a one-dir, as oppose to a one-file, based upon an answer posted at [SO Post](https://stackoverflow.com/questions/69058232/how-can-i-create-onedir-file-from-spec-file-using-pyinstaller)

The application itself works fine, and I observed that using a virtual environment and one-dir significantly decreased application start-up time in comparison to using one-file and no virtual environment. 

I've had problems with notarization due to the presence of periods in folder names and have attempted to use the "fixing script" found [here](https://github.com/pyinstaller/pyinstaller/wiki/Recipe-OSX-Code-Signing-Qt). My workflow has been as follows:

1. Create my .app, named "Main Application.app", which appears in dist, where dist is inside my desktop virtual environment folder, "virtual." 
2. Deactivate my virtual environment via typing `deactivate`. (Not familiar enough with virtual environments to know if this matters.)
3. At the Desktop directory level, use the "fixing script" by typing `Python3 fix_app_qt_folder_names_for_codesign.py 'virtual/dist/Main Application.app' `. Output states that many corrections have taken place. Here's one example:
`Fixed Desktop/virtual/dist/Main Application.app/Contents/MacOS/PyQt5/Qt5/qml/QtQuick.2`
5. Codesign using
`codesign -f -o runtime -v --deep --timestamp --entitlements /Users/fishbacp/Desktop/entitlements.plist  --timestamp -s "Developer ID Application: My Name (XXXXXX7RBW)" '/Users/fishbacp/Desktop/virtual/dist/Main Application.app'`
6. Codesigning produces no error messages, and I zip my app and upload it for notarization.

Notarization fails with the log file containing many messages such as the following:

```
      "severity": "error",
      "code": null,
      "path": "Main+Application.zip/Main Application.app/Contents/Resources/PyQt5/Qt5/qml/QtQml/Models.2/libmodelsplugin.dylib",
      "message": "The signature of the binary is invalid.",
```

In all cases, problems arise from folder names containing periods. Moreover the problematic folders are located inside the application's Contents/Resources directory and not inside Contents/MacOS. 

The fixing script described above was suggested at the end of 2019. Is there anything more up-to-date that I can use? Any suggestions in general for addressing this problem?

**Comment by rokm** [comment]
Maybe see if you can find something useful in https://github.com/pyinstaller/pyinstaller/issues/6612 ? (that was for QtWebEngine though, so it might not be the same problem and work-around as for Qml directories...)

I *think* the problem with relocating qml directories to `Contents/Resources` is that they contain binaries (.dylib files), so deep-signing the whole bundle will not sign them (because Contents/Resources is not expected to contain binaries). (Altenratively, it's the dot in the name that is preventing the signing, regardless of whether the the directories are relocated to `Contents/Resources` or not).

If you do not relocate qml directories to `Contents/Resources` but leave them at original location, and then rename each problematic qml directory to a name without a dot (e.g., rename `QtQuick.2` to `QtQuick_2`), and create symbolic link to the original name next to it (`QtQuick_2` -> `QtQuick.2`), and then re-sign the bundle, does that get you past notarization?

**Reply by rokm** [reply, marked as answer]
Also, if your application is based on QtWidget and does not use QtQuick and QML, you can probably simply remove the `PyQt5/Qt5/qml` directory from the generated app bundle and save yourself the trouble of having to deal with it.

**Reply by fishbacp** [reply]
This approach worked for me and was easiest given my lack of knowledge about creating symbolic links. Thanks.
